atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -1810,11 +1849,28 @@ private void ensureNoEmptyValues(Map<Cell, byte[]> values) {
         }
     }
 
-    private void putWritesAndLogIfTooLarge(Map<Cell, byte[]> values, SortedMap<Cell, byte[]> writes) {
+    private void putWritesAndLogIfTooLarge(
+            Map<Cell, byte[]> values,
+            ConcurrentNavigableMap<Cell, byte[]> writes,
+            Map<Cell, ChangeMetadata> metadata,
+            ConcurrentNavigableMap<Cell, ChangeMetadata> metadataForWrites) {
         for (Map.Entry<Cell, byte[]> e : values.entrySet()) {
             byte[] val = MoreObjects.firstNonNull(e.getValue(), PtBytes.EMPTY_BYTE_ARRAY);
             Cell cell = e.getKey();
-            if (writes.put(cell, val) == null) {
+            AtomicBoolean wasNull = new AtomicBoolean(false);
+            boolean hasMetadata = metadata.containsKey(cell);
+            writes.compute(cell, (k, oldVal) -> {
+                wasNull.set(oldVal == null);
+                // If we are not writing metadata for a value, we have to remove any previously stored metadata since
+                // it may not be valid for the new value.
+                if (hasMetadata) {
+                    metadataForWrites.put(cell, metadata.get(cell));
+                } else {
+                    metadataForWrites.remove(cell);
+                }
```

Comment:
This is important! One may think that we don't need to write metadata if the value does not have any associated. But what if a previous write to the same cell was done with metadata? We would be sending metadata that is only valid for an old value to TimeLock, potentially causing corruption downstream